airbase/download_api.py

```diff
@@ -0,0 +1,544 @@
+            unknown = sorted(set(countries) - COUNTRY_CODES)
+            warn(
+                f"Unknown country code(s) {', '.join(unknown)}",
+                UserWarning,
+                stacklevel=2,
+            )
+
+        payload = await self.raw_city_json(*countries)
+        cities: defaultdict[str, set[str]] = defaultdict(set)
+        for city in payload:
+            key, val = city["countryCode"], city["cityName"]
+            cities[key].add(val)
+
+        return cities
+
+    async def _download_url_to_directory(
+        self, url: str, destination: str | Path
+    ):
+        """
+        Download the content of a list of URLs ot files in a directory
+        """
+        destination = Path(destination)
+        response = await self.get(url)
+        url_path = parse.urlparse(url).path
+        filename = Path(url_path).name
+        async with aiofiles.open(destination / filename, "wb") as f:
+            await f.write(response.content)
+
+    @aiocache.cached()
+    async def _pollutant_notation_to_properties(
+        self, pollutant_notation: str
+    ) -> list[str]:
+        pollutants = await self.pollutants()
+        ids = pollutants[pollutant_notation]
+        return [
+            f"http://dd.eionet.europa.eu/vocabulary/aq/pollutant/{id}"
+            for id in ids
+        ]
+
+    async def download_downloadinfo(
+        self, *urls: DownloadInfo, destination: str | Path
+    ):
+        download_jobs = []
+
+        pbar = tqdm(
+            leave=True,
+            disable=not self.progress,
+            desc="Downloading parquets",
+            unit="files",
+            position=1,
+        )
+        pbar.set_postfix_str("Checking total number of files..")
+
+        async for parquet_url in self.parquet_file_urls(*urls):
+            task = asyncio.create_task(
+                self._download_url_to_directory(parquet_url, destination)
+            )
+
+            def update_pbar(future):
+                pbar.update(1)
+
+            task.add_done_callback(update_pbar)
+            download_jobs.append(task)
+
+        if self.progress:
+            total_files_task = asyncio.create_task(
+                self.total_num_parquet_files(*urls)
+            )
+
+            def update_pbar_total(future):
+                pbar.total = future.result()
+                pbar.set_postfix_str("")
+
+            total_files_task.add_done_callback(update_pbar_total)
+
+        for result in asyncio.as_completed(download_jobs):
+            await result
+
+    async def download(
+        self,
+        destination: str | Path,
+        countries: str | list[str] | None = None,
+        pollutants: str | list[str] | None = None,
+        cities: str | list[str] | None = None,
+        datasets: Dataset
+        | Literal["Historical", "Verified", "Unverified"]
+        | list[
+            Dataset | Literal["Historical", "Verified", "Unverified"]
+        ] = "Verified",
+    ):
```

Comment:
@avaldebe what do you think about keeping a download function like this? I think I prefer having filters rather than asking users to construct `DownloadInfo`s themselves. it matches more what they would expect from the portal.

Reply:
I was thinking that the CLI would generate the `DownloadInfo`, but I see your point.

Reply:
I want to leave it open for both CLI and Python usage. I rather have the CLI be very thin and keep the logic as close to the client as possible, and just provide different interfaces to it for convenience. That also helps prevent surprises if you switch from one to the other.